File: infrastructure/wasmbridge/validate.go
//go:build js && wasm

package wasmbridge

import "fmt"

// Bornes de sécurité, pas des constantes métier (cf. §7.2) : elles ne
// protègent pas un comportement business, seulement le moteur contre un
// message malformé ou hostile — ne jamais faire confiance à l'émetteur,
// même si le SDK/TS valide déjà côté client (§7.4).
const (
	maxKeyLength   = 1024
	maxValueLength = 1_000_000
)

func validateKey(key string) error {
	if key == "" {
		return fmt.Errorf("la clé ne peut pas être vide")
	}
	if len(key) > maxKeyLength {
		return fmt.Errorf("clé trop longue (max %d octets)", maxKeyLength)
	}
	return nil
}

func validateValue(value string) error {
	if len(value) > maxValueLength {
		return fmt.Errorf("valeur trop longue (max %d octets)", maxValueLength)
	}
	return nil
}

func validateExpireSeconds(expireSeconds int64) error {
	if expireSeconds < 0 {
		return fmt.Errorf("expireSeconds doit être positif ou nul")
	}
	return nil
}

File: web/src/client/validate.ts
// Validation à la frontière UI -> transport (§7.4) : toute donnée saisie
// dans le dashboard est vérifiée ici avant de quitter le navigateur vers le
// Worker WASM — jamais d'entrée brute non contrôlée envoyée au moteur.
// Fonctions pures, sans dépendance (pas besoin de zod pour des règles aussi
// simples : non-vide, longueur bornée, énumération). Le moteur Go revalide
// indépendamment côté Worker (infrastructure/wasmbridge/validate.go) : un
// client ne doit jamais être le seul rempart.
import type { FilterOp } from './types'

export class ValidationError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'ValidationError'
  }
}

const MAX_KEY_LENGTH = 1024
const MAX_VALUE_LENGTH = 1_000_000

const FILTER_OPS = new Set<string>(['equals', 'contains', '>', '>=', '<', '<='])

export function validateKey(key: string): void {
  if (key.length === 0) throw new ValidationError('La clé ne peut pas être vide')
  if (key.length > MAX_KEY_LENGTH) throw new ValidationError(`Clé trop longue (max ${MAX_KEY_LENGTH} caractères)`)
}

export function validateValue(value: string): void {
  if (value.length > MAX_VALUE_LENGTH) {
    throw new ValidationError(`Valeur trop longue (max ${MAX_VALUE_LENGTH} caractères)`)
  }
}

export function validateTtl(ttlSeconds: number | undefined): void {
  if (ttlSeconds !== undefined && (ttlSeconds < 0 || !Number.isFinite(ttlSeconds))) {
    throw new ValidationError('Le TTL doit être un nombre de secondes positif ou nul')
  }
}

export function validateFilterOp(op: FilterOp): void {
  if (!FILTER_OPS.has(op)) {
    throw new ValidationError(`Opérateur de filtre invalide : ${op} (attendu : ${[...FILTER_OPS].join(', ')})`)
  }
}
